### PR TITLE
Eliminate sudo chown that causes problems on systems (e.g. multipass)…

### PR DESCRIPTION
**Pull request by drbitboy**
… that do not have the xsup user, in file /setup/steps/install_milk_and_cacao.sh.

Also (i) clone a version of the cacao into milk that does not needs library lapacke, and (ii) get version 2023.2.0 of MKL

This has been made necessary by [this commit](https://github.com/milk-org/milk/commit/07f1cd4e2fd406269342ac5d88dc0bc4ccd34336) cf. [here](https://github.com/milk-org/milk/blame/dev/CMakeLists.txt#L43), where Olivier changed milk's top-level CMakeLists.txt to get the mkl-sdl (Single Dynamic Library) version of MKL, i.e. with library libmkl_rt,so*, that has all of the CBLAS and LAPACK entry points.

Also merge all of the other changes from dev.

### CODE REVIEW

**Comment by drbitboy** [comment]
This is working now, along with the ```resurrector``` branch of ```drbitboy/magao-x-config``` (=```magao-x/config```) and the ```dev-no-lapacke``` branch of ```drbitboy/milk```

File: setup/steps/install_milk_and_cacao.sh [hunk 1]
```diff
@@ -72,5 +72,6 @@ else
   make_on_data_array "cacao-${MAGAOX_ROLE,,}" /opt/MagAOX
   sudo ln -sf "/opt/MagAOX/cacao-${MAGAOX_ROLE,,}" /opt/MagAOX/cacao
 fi
+grep -q '^xsup:' /etc/passwd || exit 0
```
Comment (jaredmales) [inline setup/steps/install_milk_and_cacao.sh]: I think I'd rather have this be an if statement for future maintainability (i.e. so we could add something after without being surprised)

Comment (drbitboy) [reply setup/steps/install_milk_and_cacao.sh]: Like this?
```
if grep -q '^xsup:' /etc/passwd ; then
  ### do xsup-related stuff
fi
```

**Comment by jaredmales** [comment]
yeah, exactly

**Comment by drbitboy** [comment]
Okay, will do.

**Comment by drbitboy** [comment]
Done.

**Comment by drbitboy** [comment]
Jared, this is pretty solid now.  You should think about merging.

**Comment by drbitboy** [comment]
Jared, when the resurrector, typically induced by resuctrl, sends a USR2 signal to the xinidserver (or any) MagAOXApp, the resurrectee effectively converts it to a TERM signal by calling the SIGTERM handler for the MagAOXApp, which initiates the shutdown by assigning a 1 to the m_shutdown member of the MagAOXApp, and that causes xindiserver to shut down indiserver as well.

There is still one more thing to do (reconnect INDI drivers to a newly started INDI server), but this was the big step i.e. this [commit](https://github.com/magao-x/MagAOX/pull/131/commits/ab74384daa216aa84ed5565d93874bc92c4f457d).

**Comment by jaredmales** [comment]
Ok, got it.  Should we merge this now and do another round of tests on the instrument?

**Comment by drbitboy** [comment]
> Ok, got it. Should we merge this now and do another round of tests on the instrument?

Bottom line drbitboy recommendation:  yes, merge now after you think about any instrument code changes you or Joseph made.

**TL;DR**

Most of my changes are in the provisioning and resurrector code, and so should not affect any instrument instrument.  I did merge dev at one point because milk and cacao had moved and broken the provisioning, but merging from dev should not be an issue, right?

We would also need to merge the resurrector branch magao-x/config, although that only affect MAGAOX_ROLE=vm

I can't speak to the changes made to the instrument, so I see two choices:

1) Merge now.  For my provisioning an resurrector changes, I am fine with merging this now; it is up to you what you think about the instrument changes made here.
1.1) Risk:  the magao-x/resurrector branch is experimental AFAICT; if that is true then I don't think there is a lot of risk here as git makes it easy to reset.

2) Delay merge, Jared does instrument (MAGAOX_ROLE=aoc?) provisioning test and more instrument testing. 
2.1) I have only done MAGAOX_ROLE=VM i.e. sandbox testing
2.2) Risk:  the bigger risk, IMO, is if we delay and the drbitboy/resurrector branch ends up having conflicts with magao-x/resurrector, and then merging becomes painful.  If you can do some instrument testing quickly then this may not be an issue.

Either way, we also might want to look at merging dev into resurrector again, assuming resurrector is going to stay separate from dev for now.

**Comment by drbitboy** [comment]
P.S. the biggest problem I am having is keeping up with milk and cacao changes that break provisioning.

**Comment by joseph-long** [comment]
There is history there. We used to pin cacao to a specific commit, then we had our own fork so we could cherry-pick changes ( https://github.com/magao-x/cacao) then all the upstream repos were reorganized again and i gave up. Provisioning scripts support installing cacao on a best effort basis. The instrument itself gets more individualized attention.

**Comment by drbitboy** [comment]
> There is history there. ... Provisioning scripts support installing cacao on a best effort basis.

Okay, so this is one of those best efforts then:  it should not cause any new problems; it fixes an existing problem.

File: setup/steps/enable_vm_x11_forwarding.sh [hunk 2]
```diff
@@ -14,9 +14,9 @@ fi
 sudo systemctl restart sshd
 
 # Necessary for forwarding GUIs from the VM to the host
-source /etc/os-release # Defines $ID and $VERSION_ID so we can detect which distribution we're on
+# Defines $ID and $VERSION_ID were defined in _common.sh above, so we can detect which distribution we're on
```
Comment (joseph-long) [inline setup/steps/enable_vm_x11_forwarding.sh]: Wording wise, I would say "$ID and $VERSION_ID are sourced from /etc/os-release by _common.sh so we can detect which distribution we're on"

File: setup/steps/install_python.sh [hunk 3]
```diff
@@ -8,9 +8,15 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 source $DIR/../_common.sh
 set -euo pipefail
 
-MAMBAFORGE_VERSION="4.14.0-1"
-MAMBAFORGE_INSTALLER="Mambaforge-$MAMBAFORGE_VERSION-Linux-$(uname -p).sh"
-MAMBAFORGE_URL="https://github.com/conda-forge/miniforge/releases/download/$MAMBAFORGE_VERSION/$MAMBAFORGE_INSTALLER"
+MAMBAFORGE_VERSION="$@"
+[ -z "$MAMBAFORGE_VERSION" ] && MAMBAFORGE_VERSION="4.14.0-1"
```
Comment (joseph-long) [inline setup/steps/install_python.sh]: I'm getting us upgraded to a newer version of Mambaforge across the board so this will hit a rebase / merge conflict once https://github.com/magao-x/MagAOX/commit/37046a195b9b457003a45171ce70d7296834623b and https://github.com/magao-x/MagAOX/commit/ccd186de419ebdeaf22049090df38b18e0173172 make it into the resurrector branch. Just realized the latter hadn't been pushed yet, sorry about that.

File: setup/_common.sh [hunk 4]
```diff
@@ -128,9 +128,16 @@ function clone_or_update_and_cd() {
     # and re-enable.
 
     if [[ ! -d $parentdir/$reponame/.git ]]; then
-      echo "Cloning new copy of $orgname/$reponame"
       CLONE_DEST=/tmp/${reponame}_$(date +"%s")
-      git clone https://github.com/$orgname/$reponame.git $CLONE_DEST
+      thetopdir="$HOME/githubalt/$reponame"
```
Comment (joseph-long) [inline setup/_common.sh]: Can you explain what's going on here?

File: setup/steps/install_milk_and_cacao.sh [hunk 5]
```diff
@@ -72,5 +72,7 @@ else
   make_on_data_array "cacao-${MAGAOX_ROLE,,}" /opt/MagAOX
   sudo ln -sf "/opt/MagAOX/cacao-${MAGAOX_ROLE,,}" /opt/MagAOX/cacao
 fi
-log_info "Making /opt/MagAOX/cacao/ owned by xsup:magaox"
-sudo chown -R xsup:magaox /opt/MagAOX/cacao/
+if grep -q '^xsup:' /etc/passwd ; then
```
Comment (joseph-long) [inline setup/steps/install_milk_and_cacao.sh]: These variables should be available since this script sources `_common.sh`: https://github.com/magao-x/MagAOX/blob/dev/setup/_common.sh#L5-L18 . It should be possible to just change `sudo chown -R xsup:magaox /opt/MagAOX/cacao/` to `sudo chown -R $instrument_user:$instrument_group /opt/MagAOX/cacao/`

If I'd used `$instrument_user` and `$instrument_group` in the first place there might not have been an issue, so that's my mistake.

File: setup/_common.sh (hunk 4 above)
Comment (drbitboy) [reply setup/_common.sh]: E.g. if you
1) clone repo **https://github.com/milk-org/milk.git** to ***local*** directory **~/githubalt/milk/**, and
2) checkout branch **dev-no-lpacke** to that local directory,
then this script will use that local repo and branch as the source repo and branch, instead of the default branch dev from Github.

The same goes for any $orgname (milk-org, magao-x, etc.) and $reponame (milk, MagAOX, config, etc.).

This changes nothing if the corresponding repo sub-directory (milk/, MagAOX/, config/, etc.) is not under ~/githubalt/.

File: setup/steps/install_milk_and_cacao.sh (hunk 5 above)
Comment (drbitboy) [reply setup/steps/install_milk_and_cacao.sh]: Are you sure this is independent of xsup?

File: setup/steps/install_python.sh (hunk 3 above)
Comment (drbitboy) [reply setup/steps/install_python.sh]: > I'm getting us upgraded to a newer version of Mambaforge across the board so this will hit a rebase / merge conflict once [37046a1](https://github.com/magao-x/MagAOX/commit/37046a195b9b457003a45171ce70d7296834623b) and [ccd186d](https://github.com/magao-x/MagAOX/commit/ccd186de419ebdeaf22049090df38b18e0173172) make it into the resurrector branch. Just realized the latter hadn't been pushed yet, sorry about that.

We'll cross that bridge when we come to it.

This will affect no provisioning script that does not pass the Mambaforge version as an argument to this script.

File: setup/steps/enable_vm_x11_forwarding.sh (hunk 2 above)
Comment (drbitboy) [reply setup/steps/enable_vm_x11_forwarding.sh]: Good point, I will fix that.

File: setup/steps/install_milk_and_cacao.sh (hunk 5 above)
Comment (drbitboy) [reply setup/steps/install_milk_and_cacao.sh]: done/

File: setup/steps/enable_vm_x11_forwarding.sh (hunk 2 above)
Comment (drbitboy) [reply setup/steps/enable_vm_x11_forwarding.sh]: done.

**Comment by drbitboy** [comment]
Prefixes that now exempt resurrectees from the necessity of sending hexbeats:

- \- (dash/hyphen)
- nhb: (No HexBeat)
- py:

File: Make/common.mk
Comment (jaredmales) [inline Make/common.mk:1]: this breaks builds on the MagAO-X "flight" system, where we don't have mkl_rt.  I fixed it using local/common.mk to keep testing.  But we'll need some logic to handle this.  Except . . . we're probably moving away from MKL anyway.

File: apps/resurrector/resuctrl
Comment (jaredmales) [inline apps/resurrector/resuctrl:1]: this is very slow on MagAO-X: exao1.  Issue seems to be lsof.  Example:

```
xsup@exao1:/opt/MagAOX/source/MagAOX-resurrector/apps/resurrector$ ./resuctrl status --all
lsof: WARNING: can't stat() ext4 file system /var/lib/containers/storage/overlay
      Output information may be incomplete.
lsof: WARNING: can't stat() tmpfs file system /var/lib/containers/storage/overlay-containers/8943b7c100f0059545017e2af976aab74bf7b5edd96f6c43e2a958c2b7f60344/userdata/shm
      Output information may be incomplete.
lsof: WARNING: can't stat() overlay file system /var/lib/containers/storage/overlay/6c8e16e06ceb095b75587583bc4c7841b998d29c8d08cf2330f038f839886e8b/merged
      Output information may be incomplete.
lsof: WARNING: can't stat() ext4 file system /var/lib/containers/storage/overlay
      Output information may be incomplete.
lsof: WARNING: can't stat() tmpfs file system /var/lib/containers/storage/overlay-containers/8943b7c100f0059545017e2af976aab74bf7b5edd96f6c43e2a958c2b7f60344/userdata/shm
      Output information may be incomplete.
lsof: WARNING: can't stat() overlay file system /var/lib/containers/storage/overlay/6c8e16e06ceb095b75587583bc4c7841b998d29c8d08cf2330f038f839886e8b/merged
      Output information may be incomplete.

lsof: WARNING: can't stat() ext4 file system /var/lib/containers/storage/overlay
      Output information may be incomplete.
lsof: WARNING: can't stat() tmpfs file system /var/lib/containers/storage/overlay-containers/8943b7c100f0059545017e2af976aab74bf7b5edd96f6c43e2a958c2b7f60344/userdata/shm
      Output information may be incomplete.
lsof: WARNING: can't stat() overlay file system /var/lib/containers/storage/overlay/6c8e16e06ceb095b75587583bc4c7841b998d29c8d08cf2330f038f839886e8b/merged
      Output information may be incomplete.
^C
```
where I killed it after several seconds.   `resuctrl status XXX` doesn't seem to work.  Note that xctrl doesn't require --all on status.  --all is really just intended as safety feature on shutdown, and only shutdown.

File: apps/resurrector/HexbeatMonitor.hpp
Comment (jaredmales) [inline apps/resurrector/HexbeatMonitor.hpp:1]: The outcome of this is that the python (nhb) process is started but not connected to indiserver.  Can this be poked into indiserver without being a hexbeat process?

Comment (drbitboy) [reply apps/resurrector/HexbeatMonitor.hpp:1]: > The outcome of this is that the python (nhb) process is started but not connected to indiserver. Can this be poked into indiserver without being a hexbeat process?

Yes.  We need to write the string ```start /opt/MagAOX/drivers/fifos/<drivername>\n``` to FIFO /opt/MagAOX/drivers/fifos/indiserver.ctrl.

**TL;DR**

The way I implemented it, all processes forked by resurrector are technically hexbeat processes.  The only difference with the resurrector-forked processes that don't actually send a regular hexbeat to resurrector is that resurrector starts with a synthetic hexbeat time that will not expire.

So the question is how the process connects to the INDI server.  Formerly INDI server forked the process becasue it was in the INDI server command line, which is also how INDI server knew to find, and connect to, the process' INDI FIFOs.  Now INDI server needs to receive the message (above), sent through its control FIFO either by the process or by a third party (resurrector, or even resuctrl).  Since it is not happening, I suspect it is in MagAOXApp.hpp or more likely resurrectee.hpp.

File: apps/resurrector/resuctrl
Comment (drbitboy) [reply apps/resurrector/resuctrl:1]: /proc/ and lsof are integral to resurrector.  What operating system is exao1 running?  What does lsof have to do with these filesystems?

Can I run this exao1 setup in a VM?

Comment (drbitboy) [reply apps/resurrector/resuctrl:1]: We should be able to fix the build problem regarding mkl_rt.  Is it only the -lmkl_rt in common.mk that is causing the problem?  What was done before, when there were several -lmkl_* options in common.mk?
